core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -244,6 +242,88 @@ object Uri {
     }
   }
 
+  /** The userinfo subcomponent may consist of a user name and,
+    * optionally, scheme-specific information about how to gain
+    * authorization to access the resource.  The user information, if
+    * present, is followed by a commercial at-sign ("@") that delimits
+    * it from the host.
+    *
+    * @param username The username component, decoded.
+    *
+    * @param password The password, decoded.  Passing a password in
+    * clear text in a URI is a security risk and deprecated by RFC
+    * 3986, but preserved in this model for losslessness.
+    *
+    * @see https://www.ietf.org/rfc/rfc3986.txt#section-3.21.
+    */
+  final case class UserInfo private (username: String, password: Option[String])
```

Comment:
There are going to be more of these.  Should they continue to be nested in `Uri` or move to top level?